Tests/SwiftParserTest/Declarations.swift

```diff
@@ -375,47 +309,370 @@ final class DeclarationTests: XCTestCase {
+      { $0.parseFunctionSignature() },
+      diagnostics: [
+        DiagnosticSpec(message: "Unexpected text 'third fourth' found in function parameter")
+      ]
+    )
+  }
+
+  func testMissingClosingParenInFunctionSignature() {
+    AssertParse(
+      "(first second: Int#^DIAG^#",
+      { $0.parseFunctionSignature() },
+      diagnostics: [
+        DiagnosticSpec(message: "Expected ')' to end parameter clause")
+      ]
+    )
+  }
+
+  func testMissingOpeningParenInFunctionSignature() {
+    AssertParse(
+      "#^DIAG^#first second: Int)",
+      { $0.parseFunctionSignature() },
+      diagnostics: [
+        DiagnosticSpec(message: "Expected '(' to start parameter clause")
+      ]
+    )
+  }
+
+  func testNoParamsForFunction() {
+    AssertParse(
+      """
+      class MyClass {
+        func withoutParameters#^DIAG^#
+
+        func withParameters() {}
+      }
+      """,
+      substructure: Syntax(FunctionDeclSyntax(
+        attributes: nil,
+        modifiers: nil,
+        funcKeyword: .funcKeyword(),
+        identifier: .identifier("withoutParameters"),
+        genericParameterClause: nil,
+        signature: FunctionSignatureSyntax(
+          input: ParameterClauseSyntax(
+            leftParen: .leftParenToken(presence: .missing),
+            parameterList: FunctionParameterListSyntax([]),
+            rightParen: .rightParenToken(presence: .missing)
+          ),
+          asyncOrReasyncKeyword: nil,
+          throwsOrRethrowsKeyword: nil,
+          output: nil
+        ),
+        genericWhereClause: nil,
+        body: nil
+      )),
+      diagnostics: [
+        DiagnosticSpec(message: "Expected argument list in function declaration")
+      ]
+    )
+  }
+
+  func testThrowsInWrongLocation() {
+    AssertParse(
+      "() -> #^DIAG^#throws Int",
+      { $0.parseFunctionSignature() },
+      diagnostics: [
+        DiagnosticSpec(message: "'throws' may only occur before '->'", fixIts: ["Move 'throws' before '->'"])
+      ],
+      fixedSource: "() throws -> Int"
+    )
+  }
+
+  func testExtraneousRightBraceRecovery() {
+    // FIXME: This test case should produce a diagnostics
+    AssertParse("class ABC { let def = ghi(jkl: mno) } }")
+  }
+
+  func testMissingSubscriptReturnClause() {
+    AssertParse(
+      """
+      struct Foo {
+        subscript(x: String) #^DIAG^#{}
+      }
+      """,
+      diagnostics: [
+        DiagnosticSpec(message: "Expected '->'")
+      ]
+    )
+  }
+
+  func testClassWithLeadingNumber() {
+    AssertParse(
+      """
+      class #^DIAG^#23class {
+        // expected-error@-1 {{class name can only start with a letter or underscore, not a number}}
+        // expected-error@-2 {{'c' is not a valid digit in integer literal}}
+        func 24method() {}
+        // expected-error@-1 {{function name can only start with a letter or underscore, not a number}}
+        // expected-error@-2 {{'m' is not a valid digit in integer literal}}
```

Comment:
Were these left in to help with the FIXME?

Reply:
Yes, exactly.